### ISSUE TITLE
Dockerfile: don't use `root` as default user

### ISSUE THREAD

**Issue by curquiza** [meilisearch/MeiliSearch#2051]
Our Dockerfile to build the MeiliSearch's image uses the `root` user. We are aware this is not a good practice. However, we encountered many issues when trying to change this.

History:
- This PR https://github.com/meilisearch/MeiliSearch/pull/1759 (using the `meili` user instead of `root`) was accepted, merged, and released in v0.24.0
- We figured out this led to many issues detailed [here](https://github.com/meilisearch/MeiliSearch/issues/1969): one in the documentation (easily fixed) and one regarding Linux users.
- We tried to fix the Linux issue, without any success
- We reverted the PR and released this in v0.25.0 https://github.com/meilisearch/MeiliSearch/pull/2032

Before implementing this issue, we mean, use a non-root user for our Dockerfile, please ensure you took a look a the [whole issue conversation](https://github.com/meilisearch/MeiliSearch/issues/1969) since many things have been tried.

**Comment by K2ouMais** [comment]
We are about to go to algolia again, just because we cant run our meilisearch container as a non root user ;(

**Comment by curquiza** [comment]
Sorry for this @K2ouMais, but we did not find a way to make it work 😢

However, if you want to go back to algolia, it means you are ok with using a cloud solution instead of a self-hosted one. In this case, I recommend [our Meilisearch cloud offer](https://www.meilisearch.com/cloud?utm_campaign=oss&utm_source=github&utm_medium=meilisearch) 😊

**Comment by K2ouMais** [comment]
@curquiza No, I am not ok with it... This is a company decision... I would love to stay with Meilisearch.

**Comment by seebeen** [comment]
So - the only problem with the rootless docker is the data persistence?

@lazarivkovic - Wanna hammer out this one same as Redis Insight?

**Comment by reefland** [comment]
Just an FYI as a Melisearch noob, it runs as non-root out of the box on Kubernetes:

I randomly picked ID 568:
```
defaultPodOptions:
  securityContext:
    runAsNonRoot: true
    runAsUser: &PUID 568
    runAsGroup: &PGID 568
    fsGroup: *PGID
    fsGroupChangePolicy: OnRootMismatch
    seccompProfile: { type: RuntimeDefault }
```

And I gave it 15GB storage PVC (Persistent Volume Claim) from Ceph and it fires right up. I did my initial test using a type "EmptyDir" which is RAM backed storage (like a RAM disk) and that worked too.

shell:
```
/meili_data $ meilisearch --version
meilisearch 1.9.0

/meili_data $ df -h .
Filesystem                Size      Used Available Use% Mounted on
/dev/rbd6                14.7G    132.0K     14.6G   0% /meili_data

```

logs:
```
Config file path: "none"
Database path: "./data.ms"
Server listening on: "http://0.0.0.0:7700"
Environment: "production"
Commit SHA: "0df84bbba7db46680e83843dd6454257161a282f"
Commit date: "2024-06-27T12:47:08Z"
Package version: "1.9.0"
Anonymous telemetry: "Disabled"
A master key has been set. Requests to Meilisearch won't be authorized unless you provide an authentication key.
Check out Meilisearch Cloud! https://www.meilisearch.com/cloud?utm_campaign=oss&utm_source=engine&utm_medium=cli
Documentation: https://www.meilisearch.com/docs
Source code: https://github.com/meilisearch/meilisearch
Discord: https://discord.meilisearch.com
2024-07-31T23:19:10.759901Z INFO actix_server::builder: starting 4 workers
2024-07-31T23:19:10.759949Z INFO actix_server::server: Actix runtime found; starting in Actix runtime
2024-07-31T23:19:20.094895Z INFO HTTP request{method=GET host="10.42.2.39:7700" route=/health query_parameters= user_agent=kube-probe/1.29 status_code=200}: meilisearch: close time.busy=558µs time.idle=51.7µs
2024-07-31T23:19:20.744639Z INFO HTTP request{method=GET host="10.42.2.39:7700" route=/health query_parameters= user_agent=kube-probe/1.29 status_code=200}: meilisearch: close time.busy=70.1µs time.idle=18.3µs
2024-07-31T23:19:30.093416Z INFO HTTP request{method=GET host="10.42.2.39:7700" route=/health query_parameters= user_agent=kube-probe/1.29 status_code=200}: meilisearch: close time.busy=207µs time.idle=25.5µs
```

Hope this feedback helps.

**Comment by darkmattercoder** [comment]
Tried to use it by specifying a service user in the compose file: `user: 126:133` as I do it for all my services. However, that leads to constant

```
2025-09-23T14:02:50.681765Z ERROR meilisearch: error=Permission denied (os error 13)
Error: Permission denied (os error 13)
```
Even if the files that meili generates do have the correct user. This should work as non-root.

**Comment by ddejan75** [comment]
Same issue with me, and because of the security poicy not able to use meilisearch :(

**Comment by gypsy5oul** [comment]
@curquiza Hi, I'd like to take a shot at this, since it's still marked help wanted.

Reading #1969, the v0.24 attempt broke mainly because the data dir moved (`/data.ms` → `/home/meili/data.ms`), so existing mounts stopped matching. A version that avoids that:

- Keep `WORKDIR /meili_data` and every documented path unchanged.
- Add a small `meili` user (fixed UID/GID, e.g. 1000).
- Entrypoint: when started as root, `chown` `/meili_data` only if it's still root-owned (so upgrades from existing volumes just work), then `exec su-exec meili meilisearch "$@"` via tini.
- When started as non-root (Kubernetes `runAsNonRoot`/`fsGroup`, `docker run --user`), skip the chown and run directly.

I'd test: fresh volume, a volume created by the current root image (upgrade path), bind mounts on Linux, and a Kubernetes pod with `runAsNonRoot: true`.

Does this match what you'd accept? If so I'll open a PR. Separately, would you be interested in adding SBOM and provenance attestations next to the existing cosign signing in `publish-docker-images.yml`?
